c7n/filters/core.py

```diff
@@ -816,3 +818,107 @@ def get_resource_value(self, resource):
         if capture is None:  # regex didn't capture anything
             return None
         return capture.group(1)
+
+
+class ReduceFilter(ValueFilter):
+    """Generic reduce filter to group, sort, and limit your resources.
+
+    This example will select the longest running instance from each ASG,
+    hen randomly choose 10% of those, maxing at 15 total instances.
+
+    filters:
+      - "tag:aws:autoscaling:groupName": present
+      - type: reduce
+        group_by: "tag:aws:autoscaling:groupName"
+        sort_by: "LaunchTime"
+        order: asc
+        limit: 1
+      - type: reduce
+        order: randomize
+        limit: 15
+        limit-percent: 10
+    """
+    annotate = False
+
+    schema = {
+        'type': 'object',
+        # Doesn't mix well with inherits that extend
+        'additionalProperties': False,
+        'required': ['type'],
+        'properties': {
+            # Doesn't mix well as enum with inherits that extend
+            'type': {'enum': ['reduce']},
+            'group_by': {'type': 'string'},
+            'sort_by': {'type': 'string'},
+            'order': {'enum': ['asc', 'desc', 'reverse', 'randomize']},
+            'limit': {'type': 'number', 'minimum': 0},
+            'limit-percent': {'type': 'number', 'minimum': 0, 'maximum': 100},
+        }
+    }
+    schema_alias = True
+
+    def __call__(self, r):
```

Comment:
this method can be omitted, `process` is the method interface on a filter, __call__ is only used when delegated to by a process implementation.

Reply:
I believe this was required for how some of the tests work... I'll have to double-check.

Reply:
This also overrides the `__call__()` defined in the main ValueFilter class, which doesn't really apply to this class.

Reply:
I would just switch to a base filter class, since there's no use of inherited value filter behavior, perhaps use it as a composed for things like get_resource_value

Reply:
latest update uses a new BaseValueFilter class.